[Doc] [KubeRay] Add end-to-end tutorial for real-world RayJob workload (batch inference) (#38857)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/gcp-gke-gpu-cluster.md

# Start Google Cloud GKE Cluster with GPUs for KubeRay

> See <https://cloud.google.com/kubernetes-engine/docs/how-to/gpus> for full details, or continue reading for a quick start.

Reviewer: What does `<https://cloud.google.com/kubernetes-engine/docs/how-to/gpus>` look like after rendering?

Author: <img width="758" alt="Screenshot 2023-08-29 at 8 17 00 AM" src="https://github.com/ray-project/ray/assets/5459654/5400248b-d75f-44f1-89c8-ade09ad7f606">
My markdown linter complains if I don't add `<` and `>`

## Step 1: Create a Kubernetes cluster on GKE

> Run this command and all following commands on your local machine or on the [Google Cloud Shell](https://cloud.google.com/shell). If running from your local machine, you will need to install the [Google Cloud SDK](https://cloud.google.com/sdk/docs/install). The following command creates a Kubernetes cluster named `kuberay-gpu-cluster` with 1 CPU node in the `us-west1-b` zone. In this example, we use the `e2-standard-4` machine type, which has 4 vCPUs and 16 GB RAM.

Reviewer 2: ```suggestion
Run this command and all following commands on your local machine or on the [Google Cloud Shell](https://cloud.google.com/shell). If running from your local machine, you need to install the [Google Cloud SDK](https://cloud.google.com/sdk/docs/install). The following command creates a Kubernetes cluster named `kuberay-gpu-cluster` with 1 CPU node in the `us-west1-b` zone. This example uses the `e2-standard-4` machine type, which has 4 vCPUs and 16 GB RAM.
```

## Step 2: Create a GPU node pool

> Run the following command to create a GPU node pool for Ray GPU workers.  You can also create it from the Google Cloud Console: <https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#console>

Reviewer: ```suggestion
Run the following command to create a GPU node pool for Ray GPU workers. You can also create it from the Google Cloud Console: <https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#console>
```

```sh

> gcloud container node-pools create gpu-node-pool \

Reviewer: We can offer a variety of configuration sets. While `nvidia-l4-vws` is suitable for most examples, our EKS documentation also includes instructions for setting up a comparable GPU instance. Hence, I would recommend keeping the configuration here and provide a different configuration for the batch inference example.

```sh
gcloud container node-pools create gpu-node-pool \
  --accelerator type=nvidia-l4-vws,count=1 \
  --accelerator type=nvidia-tesla-t4,count=4,gpu-driver-version=default \
  --zone us-west1-b \
  --cluster kuberay-gpu-cluster \
  --num-nodes 1 \
  --min-nodes 0 \
  --max-nodes 1 \
  --enable-autoscaling \
  --machine-type g2-standard-4 \
  --node-taints=ray.io/node-type=worker:NoSchedule 
  --machine-type n1-standard-64 \
```

> The `--accelerator` flag specifies the type and number of GPUs for each node in the node pool. In this example, we use the [NVIDIA L4](https://cloud.google.com/compute/docs/gpus#l4-gpus) GPU. The machine type `g2-standard-4` has 1 GPU, 24 GB GPU Memory, 4 vCPUs and 16 GB RAM.

Reviewer: Because we mention `taint` in the AWS cluster setup doc, it is better to mention that how does GKE handle it.

## Step 2: Create a GPU node pool

Run the following command to create a GPU node pool for Ray GPU workers. You can also create it from the Google Cloud Console: <https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#console>

```sh
gcloud container node-pools create gpu-node-pool \
  --accelerator type=nvidia-l4-vws,count=1,gpu-driver-version=default \
  --zone us-west1-b \
  --cluster kuberay-gpu-cluster \
  --num-nodes 1 \
  --min-nodes 0 \
  --max-nodes 1 \
  --enable-autoscaling \
  --machine-type g2-standard-4 \
```

The `--accelerator` flag specifies the type and number of GPUs for each node in the node pool. In this example, we use the [NVIDIA L4](https://cloud.google.com/compute/docs/gpus#l4-gpus) GPU. The machine type `g2-standard-4` has 1 GPU, 24 GB GPU Memory, 4 vCPUs and 16 GB RAM.

.. note::

>     GKE will automatically install the GPU drivers for you.  For more details, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#create-gpu-pool-auto-drivers)

Reviewer 2: ```suggestion
    GKE automatically installs the GPU drivers for you.  For more details, see [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#create-gpu-pool-auto-drivers).
```

.. note::

>     GKE will automatically configure taints and tolerations so that only GPU pods are scheduled on GPU nodes.  For more detais, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#create)

Reviewer 2: ```suggestion
    GKE automatically configures taints and tolerations so that only GPU pods are scheduled on GPU nodes.  For more details, see [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#create)
```

```

For more details, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl).

> ## Step 4: Install NVIDIA GPU device drivers

Reviewer: Because our EKS doc mentions the NVIDIA driver, it would be better to mention how does GKE handles it.

```

> For more details, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl).

Reviewer 2: ```suggestion
For more details, see [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl).
```
